Generate anonymous objects referenced in schemas

jamietanna opened the pull request:
Closes #626, #636.

jamietanna commented on examples/authenticated-api/echo/api/api.gen.go:
```diff
@@ -288,7 +288,8 @@ type ClientWithResponsesInterface interface {
 	HTTPResponse *http.Response
-	JSON200      *[]ThingWithID
+
+	JSON200 *[]ThingWithID
```
Having a bit of difficulty tweaking the whitespace trimming in the template to avoid this diff noise 🤔

veger replied:
It is because of the if-statements you introduced. Their tempalte closing tags are `}}`, while `-}}` would prevent the new lines (I guess, didn't try it)

jamietanna commented on internal/test/components/components.gen.go:
```diff
@@ -111,9 +111,12 @@ type AdditionalPropertiesObject3 struct {
-	Inner                AdditionalPropertiesObject4_Inner `json:"inner"`
-	Name                 string                            `json:"name"`
-	AdditionalProperties map[string]interface{}            `json:"-"`
+	Inner struct {
+		Name                 string                 `json:"name"`
+		AdditionalProperties map[string]interface{} `json:"-"`
+	} `json:"inner"`
```
We shouldn't be doing this here, in the case that the object already is defined 🤔 

Looks like it may only be the `client-with-responses` that is actually not generating these correctly so I'll have a look

AlainODea commented on pkg/codegen/schema.go:
```diff
@@ -580,7 +580,16 @@ func GenFieldsFromProperties(props []Property) []string {
 		}
 
-		field += fmt.Sprintf("    %s %s", goFieldName, p.GoTypeDef())
+		if p.Schema.AdditionalTypes == nil {
+			field += fmt.Sprintf("    %s %s", goFieldName, p.GoTypeDef())
+		} else {
+			for _, td := range p.Schema.AdditionalTypes {
+				if p.Schema.TypeDecl() == td.TypeName {
+					out, _ := GenerateGoSchema(p.Schema.OAPISchema.NewRef(), nil)
+					field += fmt.Sprintf("    %s %s", goFieldName, out.GoType)
+				}
+			}
+		}
```
I'm working on the same problem in a very different way in #654. I have to admit, this is an incredibly elegant and tight solution you have here if it works. Does it work on enums?

jamietanna replied:
It doesn't appear to :grin: We'll need to use a combination of the two, methinks, to get this working

veger commented:
I am very interested to try it, when it is finished, as I think it is solving a similar issue I am experiencing... :+1:

riftsin commented:
Hi guys, I was wondering if there was any update on that issue ?

Commit:
REW!!!! Generate anonymous objects referenced in `paths`

Previously, there was no way to render any anonymous types that are
referenced in the `paths` of an OpenAPI specification.

Although there is a strong preference in the community to use the
`/components/schemas/`, there are times that it's out of our control to
move objects around.

This generates both the `requestBody` and `responses`.

TODO Closes 626 TODO

jamietanna commented:
@deepmap-marcinr I think I may need a hand with the `PascalSchema` error, currently trying to work out the intent of those tests to work out what handling needs to be added

jamietanna commented on pkg/codegen/codegen.go:
```diff
@@ -996,3 +1089,52 @@ func GetParametersImports(params map[string]*openapi3.ParameterRef) (map[string]
+					// TODO **************************************************
+					// TODO: this is where things aren't working
+					// TODO **************************************************
+					imprts, err := GoSchemaImports(mediaType.Schema)
```
Latest debugging has shown that in here, we're getting an `sref` of:

```
sref: &{#/components/schemas/pascalSchema 0x1400021a1a0}
```

Which is because at the point it's being retrieved, it's true, because it's being retrieved from inside the other spec.

Maybe this means that way that I'm pulling out all the referenced things in `GenerateTypesForPaths` is being _too_ exhaustive, and is reaching into other referenced files, which isn't ideal, as we should only be generating things from the base file

jamietanna committed:
rew!!!Generate raw types for non-JSON `requestBodies`

As part of future changes we need to be able to reference all the types
in the defined `requestBodies` of a specification.

However, as we're currently only rendering `application/json` media
types, this fails.

Instead, we can generate `json.RawMessage` (which is an alias of
`[]byte`) for these types, which allow referencing them, and
(un)marshaling data to the types.

jamietanna committed:
Generate raw types for non-JSON `responses`

As part of future changes we need to be able to reference all the types
in the defined `responses` of a specification.

However, as we're currently only rendering `application/json` media
types, this fails.

Instead, we can generate `json.RawMessage` (which is an alias of
`[]byte`) for these types, which allow referencing them, and
(un)marshaling data to the types.

jamietanna committed:
make-generate

jamietanna committed:
make-generate

jamietanna commented:
I've got this working finally :clap: 

If you're happy with it, @deepmap-marcinr, I'll sort out the Git history then merge it.

I'll test it out with a few other specs that have been mentioned to be affected, too

Commit:
Fix typo in `openapi` in spec

jamietanna commented:
Got a couple of tweaks to make:

- anonymous `requestBodies`
- anonymous `parameters`
- handle when there's i.e. `content: {}`

jamietanna committed:
rew! allow content to be empty

rew! handle absent content/mediaType

jamietanna committed:
prefactor into `generateTypesForRequestBody`

deepmap-marcinr commented on pkg/codegen/codegen.go:
```diff
@@ -462,6 +468,127 @@ func GenerateTypesForSchemas(t *template.Template, schemas map[string]*openapi3.
+					} else {
+						typeName += ToCamelCase(strings.ToLower(operationName)) + ToCamelCase(pathName)
+					}
+					typeName += fmt.Sprintf("ResponseBody%s", ToCamelCase(statusCodeOrDefault))
```
Is this sufficient to avoid type conflict? What if a request has two response bodies with different content types for the same response code? What happens in that case?

jamietanna replied:
In that case, yep, it'll handle the conflict well enough

[i.e.](https://github.com/deepmap/oapi-codegen/pull/648/files#diff-6686d60bb4293c58e363e155ee4c6b689ce7fc4ab57109d44d691a3b0099bf8bR27-R31):

```go
// NewPetResponseBody200ApplicationYaml defines model for the response body for POST /pets (200, application/yaml).
type NewPetResponseBody200ApplicationYaml = map[string]interface{}

// NewPetResponseBody200TextPlain defines model for the response body for POST /pets (200, text/plain).
type NewPetResponseBody200TextPlain = map[string]interface{}
```

mmorells commented:
Bump

massenz commented:
Is there any updated on this?
It is impacting using `go generate` for k8s `Object` DeepCopy etc, as the inlined anonymous structs cannot be consumed to generate the functions.

Simplifying the issue, we have this generated code:

```
type MyBucket struct {
	Spec *struct {
		
		ArchivePeriod *BucketSpecArchivePeriod `json:"archivePeriod,omitempty"`
	 	Cors          *struct {
	 		 ...
	 	}
	 	// other fields
}       
```
when running `go generated ./generated` we get the following error:

```
└─( go generate ./generated
command-line-arguments:-: name requested for invalid type: struct{Cors *struct{MaxAgeSeconds *float32 "json:\"maxAgeSeconds,omitempty\""; Methods *[]string "json:\"methods,omitempty\""; Origins *[]string "json:\"origins,omitempty\""; ResponseHeaders *[]string "json:\"responseHeaders,omitempty\""} "json:\"cors,omitempty\""; Status *struct{FromProvider *map[string]interface{} "json:\"fromProvider,omitempty\""} "json:\"status,omitempty\""}
Error: not all generators ran successfully
run `controller-gen object paths=types.gen.go -w` to see all available markers, or `controller-gen object paths=types.gen.go -h` for usage
generated/types.gen.go:6: running "controller-gen": exit status 1
```
By trying a few different combinations, it appears that the anonymoust `struct`s are the cause of the error.
Thanks!

gainsley commented:
Another bump for this.

Tommy-42 commented:
Hey @jamietanna do you have any news for this feature ? We would benefit a lot from this and would like to see it merge. 

Do you have any blockers ?

Cheers

jamietanna commented:
Superceded by https://github.com/oapi-codegen/oapi-codegen/pull/2366
